### ISSUE TITLE
`configure --enable-debug --enable-optimize` is not enabling optimizations

### ISSUE THREAD

**Issue by pnkfelix** [rust-lang/rust#24405]
Running `configure --enable-debug --enable-optimize` is not enabling optimizations.

Concretely, running `../configure --enable-debug --enable-optimize` yields:
```
CFG_ENABLE_DEBUG     := 1
...
CFG_DISABLE_OPTIMIZE := 1
CFG_DISABLE_OPTIMIZE_CXX := 1
...
CFG_CONFIGURE_ARGS   := --enable-debug --enable-optimize
```

And, just out of curiosity, I tried switching the order of args; that made little difference:
```
CFG_ENABLE_DEBUG     := 1
...
CFG_DISABLE_OPTIMIZE := 1
CFG_DISABLE_OPTIMIZE_CXX := 1
...
CFG_CONFIGURE_ARGS   := --enable-optimize --enable-debug
```

**Comment by pnkfelix** [comment]
(the fix for this look easy... i wonder why no one else has been flummoxed by the resulting build times though...)

**Comment by richo** [comment]
poking at a fix before I crash, mostly because I want to build rust overnight

**Comment by petrochenkov** [comment]
How much time does it take to build and test stage2 with `configure --enable-debug`. Days, I guess? I haven't tried yet.

**Comment by pnkfelix** [comment]
@petrochenkov certainly hours for me, but not a full day

**Comment by pnkfelix** [comment]
(Note that this bug is only discussing `--enable-optimize` and `--enable-debug`, and that is all that is addressed by #24408 ... but there are other flags being overridden by the relevant block of code in configure that probably have a similar problem.  I will address that separately; I want to land #24408 asap.)
